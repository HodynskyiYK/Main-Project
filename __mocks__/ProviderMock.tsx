import React from 'react'
import {MemoryRouter} from 'react-router-dom'
import { MovieModalContextProvider } from '../src/context/movie-modal-context'
import { Provider } from 'react-redux'
import { store } from '../src/store'
import { MovieDetailsContextProvider } from '../src/context/movie-details-context'

// @ts-ignore
export const ProvidersMock = ({ children }) => {
    return (
        <Provider store={store}>
            <MovieModalContextProvider>
                <MovieDetailsContextProvider>
                    <MemoryRouter>{children}</MemoryRouter>
                </MovieDetailsContextProvider>
            </MovieModalContextProvider>
        </Provider>
    )
}