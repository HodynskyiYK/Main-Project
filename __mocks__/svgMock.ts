// Mock Svg files
// https://github.com/gregberge/svgr/issues/83#issuecomment-437342861
const SvgMock = "icon-mock"
export const ReactComponent = SvgMock
export default SvgMock