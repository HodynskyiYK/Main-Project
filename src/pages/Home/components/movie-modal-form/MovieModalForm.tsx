import React, { FC, FormEvent } from 'react'
import classnames from 'classnames'
import { Input } from '../../../../components/input'
import { Select } from '../../../../components/select'
import { TextArea } from '../../../../components/textarea'
import { Button, ButtonTypes } from '../../../../components/button'
import { Modal } from '../../../../components/modal'
import { IMovieModalForm } from './movieModalFormTypes'
import styles from './MovieModalForm.module.scss'

export const MovieModalForm: FC<IMovieModalForm> = ({modalName, formValues}) => {

    const addMovieHandler = (event: FormEvent) => {
        event.preventDefault()
    }

    return (
        <Modal>
            <form
                className={styles.addMovieForm}
                onSubmit={addMovieHandler}
            >
                <p className={classnames('h1', styles.formTitle)}>{modalName}</p>
                <div className={classnames('row', 'smaller', styles.formGroup)}>
                    <div className={classnames('col', 'smaller', 'col-8')}>
                        <Input
                            inputType={'text'}
                            inputName={'movieTitle'}
                            labelText={'Title'}
                            inputPlaceholder={'Select Title'}
                            inputValue={formValues?.title}
                        />
                    </div>
                    <div className={classnames('col', 'smaller', 'col-4')}>
                        <Input
                            inputType={'date'}
                            inputName={'movieRelease'}
                            labelText={'Release date'}
                            inputPlaceholder={'Select Date'}
                            inputValue={formValues?.releaseDate}
                        />
                    </div>
                </div>
                <div className={classnames('row', 'smaller', styles.formGroup)}>
                    <div className={classnames('col', 'smaller', 'col-8')}>
                        <Input
                            inputType={'text'}
                            inputName={'movieUrl'}
                            labelText={'Movie url'}
                            inputPlaceholder={'https://'}
                            inputValue={formValues?.movieUrl}
                        />
                    </div>
                    <div className={classnames('col', 'smaller', 'col-4')}>
                        <Input
                            inputType={'text'}
                            inputName={'movieRating'}
                            labelText={'Rating'}
                            inputPlaceholder={'7.8'}
                            inputValue={formValues?.rating}
                        />
                    </div>
                </div>
                <div className={classnames('row', 'smaller', styles.formGroup)}>
                    <div className={classnames('col', 'smaller', 'col-8')}>
                        <Select
                            selectName={'movieGenre'}
                            labelText={'Genre'}
                        />
                    </div>
                    <div className={classnames('col', 'smaller', 'col-4')}>
                        <Input
                            inputType={'text'}
                            inputName={'movieRuntime'}
                            labelText={'Runtime'}
                            inputPlaceholder={'minutes'}
                            inputValue={formValues?.duration}
                        />
                    </div>
                </div>
                <div className={classnames('row', 'smaller', 'smaller', styles.formGroup)}>
                    <div className={classnames('col', 'smaller', 'col-12')}>
                        <TextArea
                            textAreaName={'TextOverview'}
                            labelText={'Overview'}
                            textAreaPlaceholder={'Movie description'}
                            textAreaValue={formValues?.overview}
                        />
                    </div>
                </div>
                <div className={styles.formButtons}>
                    <Button
                        buttonClassName={ButtonTypes.RESET_BTN}
                        buttonText={'Reset'}
                    />
                    <Button
                        buttonClassName={ButtonTypes.SEARCH_BTN}
                        buttonText={'Submit'}
                    />
                </div>
            </form>
        </Modal>
    )
}
