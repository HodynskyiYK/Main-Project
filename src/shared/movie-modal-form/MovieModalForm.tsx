import React, { FC, useState } from 'react'
import { Modal } from '../../components/modal'
import { IMovieModalForm } from './movieModalFormTypes'
import styles from './MovieModalForm.module.scss'
import classnames from 'classnames'
import { Input } from '../../components/input'
import { Select } from '../../components/select'
import { movieFormValidate, MOVIE_GENRES } from './movieModalFormConstans'
import { TextArea } from '../../components/textarea'
import { Button, ButtonTypes } from '../../components/button'
import { useAction, useTypedSelector } from '../../hooks'
import { useFormik } from 'formik'
import { IMovieItem } from '../../store/actions-types'
import { fillEmptyMovieFields } from '../../utils'
import { CompletionMoviePopup } from './CompletionMoviePopup'


export const MovieModalForm: FC<IMovieModalForm> = ({modalName, formValues}) => {
    const [addMovieSuccess, setAddMovieSuccess] = useState(false)
    const {sortBy, filterBy} = useTypedSelector(state => state.movies)
    const {addMovie, updateMovie, fetchMovies} = useAction()
    const movieForm = useFormik({
        initialValues: {
            title: formValues?.title || '',
            tagline: formValues?.tagline || '',
            vote_average: formValues?.vote_average || '',
            vote_count: formValues?.vote_count || '',
            release_date: formValues?.release_date || '',
            poster_path:  formValues?.poster_path || '',
            overview:  formValues?.overview || '',
            budget:  formValues?.budget || '',
            revenue:  formValues?.revenue || '',
            runtime:  formValues?.runtime || '',
            genres:  formValues?.genres[0] || ''
        } as IMovieItem,
        validate: movieFormValidate,
        onSubmit: async (values, {resetForm}) => {
            const movie: IMovieItem = fillEmptyMovieFields(values)
            let responseStatus
            if (formValues) {
                const updatedMovie = {
                    ...movie,
                    id: formValues.id
                }
                responseStatus = await updateMovie(updatedMovie)
            } else {
                responseStatus = await addMovie(movie)
            }
            if (+responseStatus <= 299) {
                resetForm()
                fetchMovies(filterBy, sortBy)
                setAddMovieSuccess(true)
            }
        }
    })

    return (
        <Modal>{
            !addMovieSuccess ? (
                <form
                    className={styles.addMovieForm}
                    onSubmit={movieForm.handleSubmit}
                >
                    <p className={classnames('h1', styles.formTitle)}>{modalName}</p>
                    <div className={classnames('row', 'smaller', styles.formGroup)}>
                        <div className={classnames('col', 'smaller', 'col-8')}>
                            <Input
                                inputType={'text'}
                                inputName={'title'}
                                labelText={'Title'}
                                inputPlaceholder={'Select Title'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                inputValue={movieForm.values.title}
                                errorMessage={movieForm.touched.title ? movieForm.errors.title : null}
                            />
                        </div>
                        <div className={classnames('col', 'smaller', 'col-4')}>
                            <Input
                                inputType={'date'}
                                inputName={'release_date'}
                                labelText={'Release date'}
                                inputPlaceholder={'Select Date'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                inputValue={movieForm.values.release_date}
                                errorMessage={movieForm.touched.release_date ? movieForm.errors.release_date : null}
                            />
                        </div>
                    </div>
                    <div className={classnames('row', 'smaller', styles.formGroup)}>
                        <div className={classnames('col', 'smaller', 'col-8')}>
                            <Input
                                inputType={'text'}
                                inputName={'poster_path'}
                                labelText={'Movie url'}
                                inputPlaceholder={'https://'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                inputValue={movieForm.values.poster_path}
                                errorMessage={movieForm.touched.poster_path ? movieForm.errors.poster_path : null}
                            />
                        </div>
                        <div className={classnames('col', 'smaller', 'col-4')}>
                            <Input
                                inputType={'number'}
                                inputName={'vote_average'}
                                labelText={'Rating'}
                                inputPlaceholder={'7.8'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                inputValue={movieForm.values.vote_average}
                                errorMessage={movieForm.touched.vote_average ? movieForm.errors.vote_average : null}
                            />
                        </div>
                    </div>
                    <div className={classnames('row', 'smaller', styles.formGroup)}>
                        <div className={classnames('col', 'smaller', 'col-8')}>
                            <Select
                                selectName={'genres'}
                                labelText={'Genre'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                options={MOVIE_GENRES}
                            />
                        </div>
                        <div className={classnames('col', 'smaller', 'col-4')}>
                            <Input
                                inputType={'number'}
                                inputName={'runtime'}
                                labelText={'Runtime'}
                                inputPlaceholder={'minutes'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                inputValue={movieForm.values.runtime}
                                errorMessage={movieForm.touched.runtime ? movieForm.errors.runtime : null}
                            />
                        </div>
                    </div>
                    <div className={classnames('row', 'smaller', 'smaller', styles.formGroup)}>
                        <div className={classnames('col', 'smaller', 'col-12')}>
                            <TextArea
                                textAreaName={'overview'}
                                labelText={'Overview'}
                                textAreaPlaceholder={'Movie description'}
                                onChange={movieForm.handleChange}
                                onBlur={movieForm.handleBlur}
                                textAreaValue={movieForm.values.overview}
                            />
                        </div>
                    </div>
                    <div className={styles.formButtons}>
                        <Button
                            buttonClassName={ButtonTypes.RESET_BTN}
                            buttonText={'Reset'}
                            buttonType={'reset'}
                            buttonAction={movieForm.handleReset}
                        />
                        <Button
                            buttonClassName={ButtonTypes.SEARCH_BTN}
                            buttonText={'Submit'}
                        />
                    </div>
                </form>
            ) : (
                <CompletionMoviePopup completionText={formValues ? 'The movie has been updated to database successfully' : 'The movie has been added to database successfully'}/>
            )
        }</Modal>
    )
}
