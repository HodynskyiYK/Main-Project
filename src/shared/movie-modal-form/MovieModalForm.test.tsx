import React from 'react'
import { render } from '@testing-library/react'
import { MovieModalForm } from './MovieModalForm'
import { ProvidersMock } from '../../../__mocks__/ProviderMock'

describe('MovieModalForm tests', () => {

    test('To match snapshot', () => {
        const {asFragment} = render(<ProvidersMock>
            <MovieModalForm modalName={'Form title'}/>
        </ProvidersMock>)
        expect(asFragment()).toMatchSnapshot()
    })
})