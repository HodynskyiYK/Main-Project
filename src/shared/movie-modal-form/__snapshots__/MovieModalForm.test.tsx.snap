// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieModalForm tests To match snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal"
  >
    <div
      class="modalContent"
    >
      <button
        class="closeModal"
        data-testid="close-btn"
      />
      <form
        class="addMovieForm"
      >
        <p
          class="h1 formTitle"
        >
          Form title
        </p>
        <div
          class="row smaller formGroup"
        >
          <div
            class="col smaller col-8"
          >
            <div
              class="inputField"
            >
              <label
                class="label"
                for="title"
              >
                Title
              </label>
              <input
                class="field"
                id="title"
                name="title"
                placeholder="Select Title"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col smaller col-4"
          >
            <div
              class="inputField"
            >
              <label
                class="label"
                for="release_date"
              >
                Release date
              </label>
              <input
                class="field"
                id="release_date"
                name="release_date"
                placeholder="Select Date"
                type="date"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="row smaller formGroup"
        >
          <div
            class="col smaller col-8"
          >
            <div
              class="inputField"
            >
              <label
                class="label"
                for="poster_path"
              >
                Movie url
              </label>
              <input
                class="field"
                id="poster_path"
                name="poster_path"
                placeholder="https://"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col smaller col-4"
          >
            <div
              class="inputField"
            >
              <label
                class="label"
                for="vote_average"
              >
                Rating
              </label>
              <input
                class="field"
                id="vote_average"
                name="vote_average"
                placeholder="7.8"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="row smaller formGroup"
        >
          <div
            class="col smaller col-8"
          >
            <label
              class="Label"
              for="genres"
            >
              Genre
            </label>
            <select
              class="Select"
              id="genres"
              name="genres"
            >
              <option
                value="Action"
              >
                Action
              </option>
              <option
                value="Adventure"
              >
                Adventure
              </option>
              <option
                value="Crime"
              >
                Crime
              </option>
              <option
                value="Drama"
              >
                Drama
              </option>
              <option
                value="Mystery"
              >
                Mystery
              </option>
              <option
                value="Documentary"
              >
                Documentary
              </option>
              <option
                value="Comedy"
              >
                Comedy
              </option>
              <option
                value="Romance"
              >
                Romance
              </option>
              <option
                value="Horror"
              >
                Horror
              </option>
              <option
                value="Thriller"
              >
                Thriller
              </option>
              <option
                value="Fantasy"
              >
                Fantasy
              </option>
            </select>
          </div>
          <div
            class="col smaller col-4"
          >
            <div
              class="inputField"
            >
              <label
                class="label"
                for="runtime"
              >
                Runtime
              </label>
              <input
                class="field"
                id="runtime"
                name="runtime"
                placeholder="minutes"
                type="number"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="row smaller smaller formGroup"
        >
          <div
            class="col smaller col-12"
          >
            <label
              class="Label"
              for="overview"
            >
              Overview
            </label>
            <textarea
              class="TextArea"
              cols="30"
              id="overview"
              name="overview"
              placeholder="Movie description"
              rows="7"
            />
          </div>
        </div>
        <div
          class="formButtons"
        >
          <button
            class="btn resetBtn"
            type="reset"
          >
            Reset
          </button>
          <button
            class="btn searchBtn"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
